Fixes to compile on windows using msys2

preetpalS opened the pull request:
I used these fixes to get the nmatrix gem (v0.2.1) working on Windows using mingw64 toolchain available through msys2 (https://msys2.github.io/). These fixes solve the following two issues:

- Types with names like u_int8_t not being available (this was bypassed using typedef to available types (uint8_t, uint16_t, ...).
- Ambiguous type errors since types like BYTE and INT8 are predefined on Windows (solved by using namespace qualifiers).

I am not sure if you would want to merge these changes, but the information in this pull request might be helpful to someone who is trying to use the gem on Windows (they might want to note that I already installed Lapack through msys2 package manager (pacman) using the recipe in https://github.com/Alexpux/MINGW-packages).

Note that there are issues with the g++ detection in the extconf.rb file that also prevent this gem from compiling on Windows using msys2 (mingw64) that these commits do not fix. I just bypassed the problem by bypassing the `gplusplus_version` method by making it return the version I was using (I did this on the "HackToFindG++" branch in my fork).

preetpalS commented:
When I ran `rake spec` on Windows, all tests seemed to pass as well.

`Finished in 1.01 seconds
1410 examples, 0 failures, 82 pending
C:/msys64/mingw64/bin/ruby.exe -S rspec spec/00_nmatrix_spec.rb spec/01_enum_spec.rb spec/02_slice_spec.rb spec/03_nmatrix_monkeys_spec.rb spec/blas_spec.rb spec/elementwise_spec.rb spec/homogeneous_spec.rb spec/io/fortran_format_spec.rb spec/io/harwell_boeing_spec.rb spec/io_spec.rb spec/lapack_core_spec.rb spec/math_spec.rb spec/nmatrix_yale_spec.rb spec/rspec_spec.rb spec/shortcuts_spec.rb spec/slice_set_spec.rb spec/stat_spec.rb`

v0dro commented:
Looks good. Others?

mohawkjohn commented:
These look good to me. I'm also fine with having these as two separate commits since they solve two separate but related issues.

preetpalS commented:
I took a look at my changes again, I cannot see anything wrong with the fix for the ambiguous references (where I explicitly used namespace qualifiers) but I think there is a better fix for the issue involving types with names like`u_intX_t`. If you search the repo, only `u_int8_t` is used (so the typedefs for `u_int16_t`, `u_int32_t`, and `u_int64_t` are not necessary) and it appears only twice in the same file. Since the `cstdint` header is included already and since types named like `uintX_t` are already used (including `uint8_t`), it might be better to just remove the mingw specific `typedef`s and just replace the usage of `u_int8_t` entirely with `uint8_t`. On the other hand since developers are probably only going to developing on Linux or OS X, it might be better to leave the `typedef` for mingw since the usage of types like `u_intX_t` would not trigger any warnings in the automated build and would break anyone trying to compile the gem on Windows. That's why I wasn't too sure about whether you should accept these commits.

mohawkjohn commented:
@preetpalS Yeah, I don't know why `u_int8_t` is used at all. It may have been an accident. I seem to recall using `uint8_t`, `uint16_t`, etc. throughout the rest of the code. I'd suggest doing something like:

```
#ifndef u_int8_t
typedef ...
#endif
```

and not use the MINGW ifdefs at all. Also, add a comment explaining that these should not be used.

There might also be a possibility here for creative macro use. Something that triggers a build error whenever someone uses `u_int8_t` instead of `uint8_t`, for example?

preetpalS commented:
I removed the MINGW ifdefs, replaced all uses of u_int8_t with uint8_t, and added a macro that triggers a build error whenever someone uses types like u_intX_t. My macros look like the one below:

`#define u_int8_t static_assert(false, "Please use uint8_t for cross-platform support and consistency."); uint8_t`

I did add an include statement for type_traits in the header that wasn't there before for the usage of static_assert, but that could be removed since the macros I added will only be expanded when types named like u_intX_t are actually used (but the compiler errors might not be as nice). Since these macros won't expand normally, these fixes should be fine on C++ compilers that don't support C++11 as well.

mohawkjohn commented:
@preetpalS This is great. Would you mind rebasing all of these commits? Once you do that, I'll merge. =)

preetpalS commented:
I rebased my commits off the master branch (also ended up squashing 2 of my commits and removing the one with mingw `typedef`s). There are some failures on Windows now when I test, but I believe these failures are a result of recent changes to the master branch (as the test that failed was modified in a recent commit).

`Finished in 1.42 seconds
1410 examples, 4 failures, 82 pending`

`Failed examples:`

`rspec ./spec/blas_spec.rb:187 # NMatrix::BLAS float32 exposes nrm2
rspec ./spec/blas_spec.rb:187 # NMatrix::BLAS float64 exposes nrm2
rspec ./spec/blas_spec.rb:187 # NMatrix::BLAS complex64 exposes nrm2
rspec ./spec/blas_spec.rb:187 # NMatrix::BLAS complex128 exposes nrm2
C:/msys64/mingw64/bin/ruby.exe -S rspec spec/00_nmatrix_spec.rb spec/01_enum_spe                                                                                                                     c.rb spec/02_slice_spec.rb spec/03_nmatrix_monkeys_spec.rb spec/blas_spec.rb spe                                                                                                                     c/elementwise_spec.rb spec/homogeneous_spec.rb spec/io/fortran_format_spec.rb sp                                                                                                                     ec/io/harwell_boeing_spec.rb spec/io_spec.rb spec/lapack_core_spec.rb spec/math_                                                                                                                     spec.rb spec/nmatrix_yale_spec.rb spec/rspec_spec.rb spec/shortcuts_spec.rb spec                                                                                                                     /slice_set_spec.rb spec/stat_spec.rb failed
C:/msys64/mingw64/bin/ruby.exe -S rspec spec/00_nmatrix_spec.rb spec/01_enum_spe                                                                                                                     c.rb spec/02_slice_spec.rb spec/03_nmatrix_monkeys_spec.rb spec/blas_spec.rb spe                                                                                                                     c/elementwise_spec.rb spec/homogeneous_spec.rb spec/io/fortran_format_spec.rb sp                                                                                                                     ec/io/harwell_boeing_spec.rb spec/io_spec.rb spec/lapack_core_spec.rb spec/math_                                                                                                                     spec.rb spec/nmatrix_yale_spec.rb spec/rspec_spec.rb spec/shortcuts_spec.rb spec                                                                                                                     /slice_set_spec.rb spec/stat_spec.rb`

mohawkjohn commented:
Argh, that's frustrating. Maybe #491. @arafatk can you please have a look at this?

mohawkjohn commented:
I'm going to go ahead and merge since it improves things in Windows overall, but I'm not going to close this ticket.

Arafatk commented:
@mohawkjohn @preetpalS I am sorry, for the issue. 
The issue is because of this [pr](https://github.com/SciRuby/nmatrix/pull/491/files) where I added a check to skip division by zero, I know that their are a lot of precision issues in c, but this worked fine on many test cases, so I sent the pr. 
I am not sure, how to go about solving this, because I don't have windows and I could use windows on my friends laptop but their are a lot of proxy issues in my university that would take me quite a lot of time to even get the basic setup running. The originally file in FORTRAN is implemented [here](http://www.netlib.org/lapack/explore-html/da/d7f/dnrm2_8f_source.html) read line 94 which checks for zero. 

@preetpalS It would be a huge favor if you could look into the issue.

Again, I am very sorry for the inconvenience. If you guys are busy i will look over it on the weekend.

preetpalS commented:
Thanks for merging this pull request =).

I will take a look into the issue.

preetpalS commented:
@arafatk There's no need to apologize, I made the mistake during the testing process. I did not recompile the library after checking out the latest version of the library's repository. The test suite still passes (see https://github.com/SciRuby/nmatrix/issues/495).
